test/e2e/specs/editor/various/undo.spec.js

```diff
@@ -0,0 +1,516 @@
+/**
+ * WordPress dependencies
+ */
+const { test, expect } = require( '@wordpress/e2e-test-utils-playwright' );
+
+test.use( {
+	undoUtils: async ( { page }, use ) => {
+		await use( new UndoUtils( { page } ) );
+	},
+} );
+
+class UndoUtils {
+	constructor( { page } ) {
+		this.page = page;
+
+		this.getSelection = this.getSelection.bind( this );
+	}
+
+	async getSelection() {
+		return await this.page.evaluate( () => {
+			const selectedBlock = document.activeElement.closest( '.wp-block' );
+			const blocks = Array.from(
+				document.querySelectorAll( '.wp-block' )
+			);
+			const blockIndex = blocks.indexOf( selectedBlock );
+
+			if ( blockIndex === -1 ) {
+				return {};
+			}
+
+			let editables;
+
+			if ( selectedBlock.getAttribute( 'contenteditable' ) ) {
+				editables = [ selectedBlock ];
+			} else {
+				editables = Array.from(
+					selectedBlock.querySelectorAll( '[contenteditable]' )
+				);
+			}
+
+			const editableIndex = editables.indexOf( document.activeElement );
+			const selection = window.getSelection();
+
+			if ( editableIndex === -1 || ! selection.rangeCount ) {
+				return { blockIndex };
+			}
+
+			const range = selection.getRangeAt( 0 );
+			const cloneStart = range.cloneRange();
+			const cloneEnd = range.cloneRange();
+
+			cloneStart.setStart( document.activeElement, 0 );
+			cloneEnd.setStart( document.activeElement, 0 );
+
+			/**
+			 * Zero width non-breaking space, used as padding in the editable DOM
+			 * tree when it is empty otherwise.
+			 */
+			const ZWNBSP = '\ufeff';
+
+			return {
+				blockIndex,
+				editableIndex,
+				startOffset: cloneStart.toString().replace( ZWNBSP, '' ).length,
+				endOffset: cloneEnd.toString().replace( ZWNBSP, '' ).length,
+			};
+		} );
+	}
+}
+
+test.describe( 'undo', () => {
+	test.beforeEach( async ( { admin } ) => {
+		await admin.createNewPost();
+	} );
+
+	test( 'should undo typing after a pause', async ( {
+		editor,
+		page,
+		pageUtils,
+		undoUtils,
+	} ) => {
+		await page.click( 'role=button[name="Add default block"i]' );
+		await page.keyboard.type( 'before pause' );
+		await new Promise( ( resolve ) => setTimeout( resolve, 1000 ) );
+		await page.keyboard.type( ' after pause' );
+
+		const after = await editor.getEditedPostContent();
+
+		expect( after ).toBe(
+			`<!-- wp:paragraph -->
+<p>before pause after pause</p>
+<!-- /wp:paragraph -->`
+		);
```

Comment:
Nit: `editor.getBlocks` could be more readable, see https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/e2e/overusing-snapshots.md.

```suggestion
		await expect.poll( editor.getBlocks ).toMatchObject( [ {
			name: 'core/paragraph',
			attributes: { content: 'before pause after pause' },
		} ] );
```

Same for other assertions below.